### PR TITLE
Adds Detective's SpyGlasses and related requirements

### PR DESCRIPTION
**Pull request by bloons3**
## About The Pull Request
Give the Detective some nice gimmick items to use instead of SecHUDs

The Detective's SpyGlasses can be worn and offer a remote view of a camera, which can be attached to clothing.

They are flash resistant, like regular sunglasses, and start in the Detective's Wardrobe.

- [x] https://github.com/tgstation/tgstation/pull/47241 - Adds generic popup component
- [x]  tgstation/tgstation#46891 - Add SpyGlasses
- [x]  tgstation/tgstation#56883 - Turn the camera into an accessory
- [x]  tgstation/tgstation#53355 - Merge the Det's Sunglasses into their Spy Glasses
Related PRs:

https://github.com/tgstation/tgstation/pull/46891
https://github.com/tgstation/tgstation/pull/49960
https://github.com/tgstation/tgstation/pull/50669
https://github.com/tgstation/tgstation/pull/51182
https://github.com/tgstation/tgstation/pull/52728
https://github.com/tgstation/tgstation/pull/53355
https://github.com/tgstation/tgstation/pull/54403
https://github.com/tgstation/tgstation/pull/56883

## Why It's Good For The Game

image

image

## Changelog
:cl: FatCat1978, ninjanomnom, stylemistake, nemvar, Timberpoes, Fikou, TemporalOroboros
add: The Detective now features a classy and stylish piece of spy equipment. The SpyGlasses offer a remote view of  a linked camera, which can be inconspicuously worn on clothing
fix: Makes it possible to wear the spy bug by making it actually an accessory clothing item.
code: Moves out the recursive movement tracker found in the orbiting components to a datum usable by other things
/:cl:

### CODE REVIEW

**Comment by bloons3** [comment]
BRUH
image

File: code/game/objects/items/devices/spyglasses.dm [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+/obj/item/clothing/glasses/sunglasses/spy
+	desc = "Made by Nerd. Co's infiltration and surveillance department. Upon closer inspection, there's a small screen in each lens."
+	actions_types = list(/datum/action/item_action/activate_remote_view)
+	var/obj/item/clothing/accessory/spy_bug/linked_bug
+
+/obj/item/clothing/glasses/sunglasses/spy/proc/show_to_user(mob/user)//this is the meat of it. most of the map_popup usage is in this.
+	if(!user)
+		return
+	if(!user.client)
+		return
```
Comment (Archanial) [inline code/game/objects/items/devices/spyglasses.dm]:
```suggestion
	if(!user?.client)
		return
```

File: code/game/objects/items/devices/spyglasses.dm [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+/obj/item/clothing/glasses/sunglasses/spy/dropped(mob/user)
+	. = ..()
+	user.client.close_popup("spypopup")
+
+/obj/item/clothing/glasses/sunglasses/spy/ui_action_click(mob/user)
+	show_to_user(user)
+
+/obj/item/clothing/glasses/sunglasses/spy/item_action_slot_check(slot)
+	if(slot == ITEM_SLOT_EYES)
+		return TRUE
```
Comment (utilbird) [inline code/game/objects/items/devices/spyglasses.dm]:
```suggestion
	return slot == ITEM_SLOT_EYES
```

File: code/game/objects/items/devices/spyglasses.dm [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+/obj/item/clothing/glasses/sunglasses/spy/Destroy()
+	if(linked_bug)
+		linked_bug.linked_glasses = null
+	. = ..()
```
Comment (utilbird) [inline code/game/objects/items/devices/spyglasses.dm]: `.` is the return value of a proc so the direct return method is preferred at the end of a proc
```suggestion
	return ..()
```

File: code/game/objects/items/devices/spyglasses.dm [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		instance.screen_loc = "spypopup_map:CENTER"
+		cam_plane_masters += instance
+
+/obj/item/clothing/accessory/spy_bug/Destroy()
+	if(linked_glasses)
+		linked_glasses.linked_bug = null
+	qdel(cam_screen)
+	QDEL_LIST(cam_plane_masters)
+	qdel(tracker)
+	. = ..()
```
Comment (utilbird) [inline code/game/objects/items/devices/spyglasses.dm]:
```suggestion
	return ..()
```

File: code/game/objects/items/devices/spyglasses.dm [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+//it needs to be linked, hence a kit.
+/obj/item/storage/box/rxglasses/spyglasskit
+	name = "spyglass kit"
+	desc = "this box contains <i>cool</i> nerd glasses; with built-in displays to view a linked camera."
+
+/obj/item/paper/fluff/nerddocs
+	name = "Espionage For Dummies"
+	color = "#FFFF00"
+	desc = "An eye gougingly yellow pamphlet with a badly designed image of a detective on it. the subtext says \" The Latest way to violate privacy guidelines!\" "
```
Comment (utilbird) [inline code/game/objects/items/devices/spyglasses.dm]: weewoo
```suggestion
	desc = "An eye gougingly yellow pamphlet with a badly designed image of a detective on it. the subtext says \" The Latest Way To Violate Privacy Guidelines!\" "
```

File: code/game/objects/items/devices/spyglasses.dm (hunk 2 above)
Comment (bloons3) [reply code/game/objects/items/devices/spyglasses.dm]: Personal preference?

Comment (francinum) [reply code/game/objects/items/devices/spyglasses.dm]: branch bad

**Comment by DatBoiTim** [comment]
Based?

**Comment by PowerfulBacon** [comment]
The balance of these seem pretty off, a remote camera that you can view anywhere (not even at a security camera console) and in a second window 24/7.
It would be better as something like the remote camera but you need a console in detectives office to view it, or it takes pictures at the camera with a short cooldown.

**Review by PowerfulBacon — DISMISSED** [review]
Unintentional changes to skin

**Comment by bloons3** [comment]
> Unintentional changes to skin

Which skin? Not seeing any icon changes in the git diff

**Review by utilbird — COMMENTED** [review]
`ITEM_SLOT_EYES` is a bitflag and it'll need to use operators as such. `&` instead of `==`, etc.

**Comment by utilbird** [comment]
Now that I look at it, a lot of our code still uses equality checks. I'll bring it up to date sometime soon:tm:

File: code/game/objects/items/devices/spyglasses.dm [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+		return
+	user.client.setup_popup("spypopup", 3, 3, 2)
+	user.client.register_map_obj(linked_bug.cam_screen)
+	for(var/plane in linked_bug.cam_plane_masters)
+		user.client.register_map_obj(plane)
+	linked_bug.update_view()
+
+/obj/item/clothing/glasses/sunglasses/spy/equipped(mob/user, slot)
+	. = ..()
+	if(slot != ITEM_SLOT_EYES)
```
Comment (utilbird) [inline code/game/objects/items/devices/spyglasses.dm]:
```suggestion
	if(!(slot & ITEM_SLOT_EYES))
```

File: code/game/objects/items/devices/spyglasses.dm [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+/obj/item/clothing/glasses/sunglasses/spy/dropped(mob/user)
+	. = ..()
+	user.client.close_popup("spypopup")
+
+/obj/item/clothing/glasses/sunglasses/spy/ui_action_click(mob/user)
+	show_to_user(user)
+
+/obj/item/clothing/glasses/sunglasses/spy/item_action_slot_check(slot)
+	return slot == ITEM_SLOT_EYES
```
Comment (utilbird) [inline code/game/objects/items/devices/spyglasses.dm]:
```suggestion
	return slot & ITEM_SLOT_EYES
```

**Review by utilbird — APPROVED** [review]

**Pull request merged**